example/advanced_advection/advanced_advection_package.cpp

```diff
@@ -0,0 +1,523 @@
+  Real ang_3 = pin->GetOrAddReal("Advection", "ang_3", -999.9);
+
+  Real ang_2_vert = pin->GetOrAddBoolean("Advection", "ang_2_vert", false);
+  Real ang_3_vert = pin->GetOrAddBoolean("Advection", "ang_3_vert", false);
+
+  // For wavevector along coordinate axes, set desired values of ang_2/ang_3.
+  //    For example, for 1D problem use ang_2 = ang_3 = 0.0
+  //    For wavevector along grid diagonal, do not input values for ang_2/ang_3.
+  // Code below will automatically calculate these imposing periodicity and exactly one
+  // wavelength along each grid direction
+  Real x1size = pin->GetOrAddReal("parthenon/mesh", "x1max", 1.5) -
+                pin->GetOrAddReal("parthenon/mesh", "x1min", -1.5);
+  Real x2size = pin->GetOrAddReal("parthenon/mesh", "x2max", 1.0) -
+                pin->GetOrAddReal("parthenon/mesh", "x2min", -1.0);
+  Real x3size = pin->GetOrAddReal("parthenon/mesh", "x3max", 1.0) -
+                pin->GetOrAddReal("parthenon/mesh", "x3min", -1.0);
+
+  // User should never input -999.9 in angles
+  if (ang_3 == -999.9) ang_3 = std::atan(x1size / x2size);
+  Real sin_a3 = std::sin(ang_3);
+  Real cos_a3 = std::cos(ang_3);
+
+  // Override ang_3 input and hardcode vertical (along x2 axis) wavevector
+  if (ang_3_vert) {
+    sin_a3 = 1.0;
+    cos_a3 = 0.0;
+    ang_3 = 0.5 * M_PI;
+  }
+
+  if (ang_2 == -999.9)
+    ang_2 = std::atan(0.5 * (x1size * cos_a3 + x2size * sin_a3) / x3size);
+  Real sin_a2 = std::sin(ang_2);
+  Real cos_a2 = std::cos(ang_2);
+
+  // Override ang_2 input and hardcode vertical (along x3 axis) wavevector
+  if (ang_2_vert) {
+    sin_a2 = 1.0;
+    cos_a2 = 0.0;
+    ang_2 = 0.5 * M_PI;
+  }
+
+  Real x1 = x1size * cos_a2 * cos_a3;
+  Real x2 = x2size * cos_a2 * sin_a3;
+  Real x3 = x3size * sin_a2;
+
+  // For lambda choose the smaller of the 3
+  Real lambda = x1;
+  if ((pin->GetOrAddInteger("parthenon/mesh", "nx2", 1) > 1) && ang_3 != 0.0)
+    lambda = std::min(lambda, x2);
+  if ((pin->GetOrAddInteger("parthenon/mesh", "nx3", 1) > 1) && ang_2 != 0.0)
+    lambda = std::min(lambda, x3);
+
+  // If cos_a2 or cos_a3 = 0, need to override lambda
+  if (ang_3_vert) lambda = x2;
+  if (ang_2_vert) lambda = x3;
+
+  // Initialize k_parallel
+  Real k_par = 2.0 * (M_PI) / lambda;
+
+  pkg->AddParam<>("amp", amp);
+  pkg->AddParam<>("vel", vel);
+  pkg->AddParam<>("vx", vx);
+  pkg->AddParam<>("vy", vy);
+  pkg->AddParam<>("vz", vz);
+  pkg->AddParam<>("k_par", k_par);
+  pkg->AddParam<>("cos_a2", cos_a2);
+  pkg->AddParam<>("cos_a3", cos_a3);
+  pkg->AddParam<>("sin_a2", sin_a2);
+  pkg->AddParam<>("sin_a3", sin_a3);
+
+  // initialize the alias tables to sample the number of iterations from a discrete power
+  // law distributions, for description of the Alias Method, see
+  // https://en.wikipedia.org/wiki/Alias_method
+  // https://www.keithschwarz.com/darts-dice-coins/
+  // https://gist.github.com/Liam0205/0b5786e9bfc73e75eb8180b5400cd1f8
```

Comment:
Excellent question. I'm always for using stdlib and not reinventing (or re-implementing) the wheel. I did look at `std::discrete_distribution` and considered using it, but eventually decided to implement the alias method myself for several reasons:

1. It's not clear to me which method `std::discrete_distribution` uses internally. It says it's `operator()` method is amortized constant, so it could be less efficient than the alias method, which is guaranteed constant.
2. It's also not clear to me if `std::discrete_distribution` keeps any dynamic internal state and thus if it is thread-safe, which we would need, otherwise we'd have to make many copies of it, one for each thread.
3. Most importantly, since we sample the number of iterations on the device, we would need `std::discrete_distribution` to work on the GPU in parallel. I didn't try it and it's possible that it works, but I decided I'd rather know exactly what's happening on the GPU and how much memory we use. With the alias method implemented ourselves, we just read from two small `Kokkos::view`s (the probability and alias tables) of which we only need one copy and the actual sampling just involves two table look ups without any dynamic state, so we know it's thread-safe (given that Kokkos' random number generator is thread-safe) and lightweight on the device.

It may be interesting to try to use `std::discrete_distribution` on the GPU, but I suspect we will get some additional UVM transfers because we can't tell `std::discrete_distribution` to allocate its internal state on the device, as we can with `Kokkos::view`s.

Reply:
I see. Interesting.

Yeah I don't think `std::discrete_distribution` would work on GPU. I think the strategy there would be to call it on CPU and precompute random numbers, and then fill a kokkos view with them.

Would it be worth saving this alias method to a `parthenon` utility so it's available in the future and to downstream codes?

Reply:
I think that's a good idea. I'll split it off.